Add windowed(): Sliding window over an input range

Kijewski opened the pull request:
E.g. for a range `(1,2,3,4,5)` a window of 2 elements is `[(1,2), (2,3), (3,4), (5)]`. For a range  `(1,2,3,4,5,6)` a window of 2 elements with a stepsize of 3 is `[(1,2), (4,5)]`.

Inspired by [more_itertools.windowed()][1].

  [1]: https://more-itertools.readthedocs.io/en/v7.2.0/api.html#more_itertools.windowed

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1655,6 +1656,107 @@ struct reverse {
+        RX_ASSERT(window_size > 0);
+        RX_ASSERT(step_size > 0);
+    }
+    constexpr windowed(const windowed&);
```
You are missing the definition of it. Though, I'm confused why you even declare them as the implicitly declared copy/move constructors/assignment operators are already constexpr and noexcept.

Kijewski replied:
Thank you for reviewing my code! If I have a user defined constructor, the implicit constructors aren't automatically generated, are they? I'll add the missing `= default`.

Hazurl replied:
They are still implicitly defined: https://en.cppreference.com/w/cpp/language/copy_constructor#Deleted_implicitly-declared_copy_constructor

Kijewski replied:
Thank you for the link! Lazy as I am I never really bothered to remember when I *have to* add the c'tors. I'll simply always add all five lines:

```c++
constexpr T() noexcept = default;
constexpr T(const T&) noexcept = default;
constexpr T(T&&) noexcept = default;
constexpr T &operator =(const T&) noexcept = default;
constexpr T &operator =(T&&) noexcept = default;
```

Hazurl commented on include/rx/ranges.hpp:
```diff
@@ -1655,6 +1656,107 @@ struct reverse {
 
+/// Sliding window over an input range.
+///
+/// E.g. for a range `(1,2,3,4,5)` a window of 2 elements is `[(1,2), (2,3), (3,4), (5)]`.
```
do you mean `[(1,2), (2,3), (3,4), (4,5)]` ?

Kijewski replied:
Actually, no. Otherwise the output would be the same for an input `(1,2,3,4,5)` and `(1,2,3,4,4,5)`.

I was not sure how to handle "underfull" groups, whether to use empty optional<>s, a padding element, or let the user decide. Returning a range with too fewer elements is the simplest solution, so I chose to use that. Maybe the other options could be added later.

Hazurl replied:
with `(1,2,3,4,4,5)` I would expect `(1,2),(2,3),(3,4),(4,4),(4,5)` 🤔.

Kijewski replied:
Argh. I'm dumb, and I've confused myself. :) Sorry, I'll fix the description.

Hazurl reviewed (commented):
I think you could return a view instead of a forward_list. Exactly like https://github.com/simonask/rx-ranges/pull/6. `next` would just call `step_size_` times `input_.next()` and `get` would returns `input_ | take(window_size_)`.

Kijewski commented:
@Hazurl, you are right. I only had to idea to return ranges after I wrote this PR. I'll adapt it.

Kijewski commented:
> I think you could return a view instead of a forward_list. Exactly like #6. `next` would just call `step_size_` times `input_.next()` and `get` would returns `input_ | take(window_size_)`.

@Hazurl, I did not come up with another that

- does not require some sort of deque or
- iterate over the input range multiple times or
- slides out of the window: (1,2,3,4) -> window over 3 elements ->  [(1,2,3), (2,3,4), (3,4), (4)].

Hazurl commented:
I made a PR to your branch (using the `intput_ | take(window_size_)` option), but now I see what you mean by `iterate other the input range multiple times`. `get` is called multiple times per element if `window_size < step_size`.
I'm not sure what's the best decision to this, probably `std::deque` as you suggest.

Kijewski committed:
Add: windowed(), sliding window over a range

E.g. for a range `(1,2,3,4,5)` a window of 2 elements is
`[(1,2), (2,3), (3,4), (5)]`. For a range `(1,2,3,4,5,6)` a window of 2
elements with a stepsize of 3 is `[(1,2), (4,5)]`.

Inspired by [more_itertools.windowed()][1].

[1]: https://more-itertools.readthedocs.io/en/v7.2.0/api.html#more_itertools.windowed

simonask commented:
I think the reliance on `std::deque` is unfortunate, since it would now be the only place in the library where internal storage of more than one element is allocated. I might give it a shot to see if it isn't possible to implement in terms of copied subranges, like we do in a couple of other places.

`std::deque` is especially problematic (but I completely understand the rationale for choosing it), because it has a surprising defect on MSVC, where [the internal contiguous block size is very small (at most 16 bytes!)](https://github.com/microsoft/STL/blob/master/stl/inc/deque#L27). This means that it effectively performs like a ring buffer of `std::vector<std::unique_ptr<T>>`. In other words, it is completely arguably broken on MSVC. :-)

Hazurl commented:
There's another possibility to make this work. Essentially keeping a vector of `window_size` elements. and returning the range `element | cycle() | skip(start_index) | take(window_size)` where `start_index` is incremented by `step_size` in the `next` method.
So no memory overhead, one allocation in the constructor and continuous elements. For performance reason, we could make a custom range to replace `element | cycle() | skip(start_index) | take(window_size)` (unless the compiler is smart enough).
Also if `step_size > window_size`, `get` just returns `as_input_range(elements)`.

Kijewski commented:
One can re-implement a deque in various ways, esp. if its maximum size is known. My first implementation used a forward_list, I could implement a ring buffer using std::vector, etc., but any naive implementation  using std::vector would needlessly turn this into an O(N²) algorithm, or am I misunderstanding what you propose?

Hazurl commented:
I don't know where you get that `O(n²)` 🤔 
In my solution, `get` would be `O(1)` as it just creates a range. And `next` would store the next `window_size` elements (`window_size - step_size` if `step_size < window_size`). The inner range is iterated once, so this solution is `O(n)`, right?

Kijewski commented:
Ah, but that would reintroduce the "sliding out of the window" problem: https://github.com/Kijewski/rx-ranges/pull/1#discussion_r347101087

Hazurl commented:
So I tested: https://github.com/Kijewski/rx-ranges/pull/2
It passes all the tests from the last version with `std::deque`, so I don't think it slides out of the window.

simonask commented:
I'm still not convinced of the necessity of internal heap-allocated storage. I think it is a surprising side-effect to users.

Could the same thing not be achieved by something similar to the this:

(details left out for brevity)

```
size_t step_by;
size_t window_size;
Inner inner;
RX_OPTIONAL<...> storage;

auto get() {
    return *storage;
}
auto next() {
    advance_by(inner, step_by);
    storage.emplace(inner | in_groups_of_exactly(window_size) | take(1));
}
```

simonask commented:
Oh, and just in case this isn't clear: I have to say I am strongly against shipping customized implementations of `std::deque` or implementations of ring buffers on top of `std::vector` as part of this library, if that is what anyone is suggesting. ;-)

Hazurl commented:
We used `take(window_size)` at some point which has the same problem as `in_groups_of_exactly`, it will in the case of non-idempotent range, call `get` multiple time per element (if `step_size < window_size`).
The "ring buffer" implementation is just an index for which element is the start of the range. 

Unfortunately, if we can't find a good solution, we may just drop this range...

simonask commented:
Ah, I see, and it's not trivial to use `as_idempotent_input_range()`, because of the need to "rewind" when advancing by less than the window size.

However, I do suspect that the vast majority of ranges are actually idempotent in practice. It might make sense to specialize `windowed()` for idempotent and non-idempotent cases, where only the non-idempotent case allocates memory proportional to the window size. The easiest way to formulate that might be to produce the window in the same way, but materialize it with `to_vector()` in the non-idempotent case.

Hazurl commented:
That could be better

Hazurl commented:
I started to implement it, but I run into a problem. I returns `input | take(window_size)` in `get`. But to correctly implement the behavior I must check the size of the returned range so that the latter is not incomplete. Which means it's `O(n²)`. That's not good.
I see 2 solutions:
- Implement an optional method `exact_size` which in the worst case is `O(n)` but can be `O(1)` for some ranges.
- Change the behavior to output all ranges of size `window_size` with a certain step. Even when it's incomplete.

I would be in favor of the second. Though the method `exact_size` could be a nice feature.

Kijewski commented:
We might postpone the addition of this function. I'm feeling lazy this evening, sorry. Let's release v2.0 first.
